env/env_encrypt2.cc

```diff
@@ -0,0 +1,558 @@
+    data_ptr = (uint64_t*)data;
+    *data_ptr ^= block_out.nums[0];
+    data_ptr = (uint64_t*)(data + 8);
+    *data_ptr ^= block_out.nums[1];
+  } else {
+    status = Status::NotSupported(
+        "libcrypto not available for encryption/decryption.");
+  }
+
+  return status;
+}
+
+Status AESBlockAccessCipherStream::DecryptBlock(uint64_t blockIndex, char* data,
+                                                char* scratch) {
+  return EncryptBlock(blockIndex, data, scratch);
+}
+
+Status CTREncryptionProvider2::CreateNewPrefix(const std::string& /*fname*/,
+                                               char* prefix,
+                                               size_t prefixLength) {
+  Status s;
+  if (EncryptedEnv2::crypto_.IsValid()) {
+    if (sizeof(Prefix0_t) <= prefixLength) {
+      int ret_val;
+
+      Prefix0_t* pf = {(Prefix0_t*)prefix};
+      memcpy(pf->key_description_, key_desc_.desc, sizeof(key_desc_.desc));
+      ret_val = EncryptedEnv2::crypto_.RAND_bytes(
+          (unsigned char*)&pf->nonce_,
+          AES_BLOCK_SIZE / 2);  // RAND_poll() to initialize
+      if (1 != ret_val) {
+        s = Status::NotSupported("RAND_bytes failed");
+      }
+    } else {
+      s = Status::NotSupported("Prefix size needs to be 28 or more");
+    }
+  } else {
+    s = Status::NotSupported("RAND_bytes() from libcrypto not available.");
+  }
+
+  return s;
+}
+
+// Returns an Env that encrypts data when stored on disk and decrypts data when
+// read from disk.
+Env* NewEncryptedEnv2(Env* base_env, EncryptedEnv2::ReadKeys_t encrypt_read,
+                      EncryptedEnv2::WriteKey_t encrypt_write) {
+  Env* ret_env{base_env};
+  EncryptedEnv2* new_env{nullptr};
+
+  if (Env::Default() == base_env) {
+    // use safer static construction so libcrypto is synchronously loaded
+    new_env =
+        (EncryptedEnv2*)EncryptedEnv2::Default(encrypt_read, encrypt_write);
+  } else if (nullptr != base_env) {
+    new_env = new EncryptedEnv2(base_env, encrypt_read, encrypt_write);
+  }
+
+  // warning, dynamic loading of libcrypto could be delayed ... making this
+  // false
+  if (nullptr != new_env && new_env->IsValid()) {
+    ret_env = new_env;
+  }
+
+  return ret_env;
+}
+
+EncryptedEnv2::EncryptedEnv2(Env* base_env,
+                             EncryptedEnv2::ReadKeys_t encrypt_read,
+                             EncryptedEnv2::WriteKey_t encrypt_write)
+    : EnvWrapper(base_env),
+      encrypt_read_(encrypt_read),
+      encrypt_write_(encrypt_write),
+      valid_(false) {
+  valid_ = crypto_.IsValid();
+
+  // warning, dynamic loading of libcrypto could be delayed ... making this
+  // false
+  if (IsValid()) {
+    crypto_.RAND_poll();
+  }
+}
+
+EncryptedEnv2::EncryptedEnv2(Env* base_env)
+    : EnvWrapper(base_env), valid_(false) {}
+
+void EncryptedEnv2::SetKeys(EncryptedEnv2::ReadKeys_t encrypt_read,
+                            EncryptedEnv2::WriteKey_t encrypt_write) {
+  encrypt_read_ = encrypt_read;
```

Comment:
what if there's concurrent access to the keys when we call `SetKeys`? Do we need a mutex?

Reply:
This routine is currently called only by Encrypt2::Default().  There is not even a unit test that calls it.  How about making it "protected:" instead.  If we put a mutex, that implies every access to the keys would also need to lock the same mutex.  That is more overhead than I would really like for an extremely unlikely use case.

Reply:
I implemented the move to "protected" and added comment concerning not thread safe.

Reply:
The use case is online key rotation: caller want to replace the write key without interrupting service.

Reply:
ok, but that implies changing on file open boundaries ... so more code would be require to save current write key within an open writable file.  I would prefer to leave that as an extension, a new PR, unless you say that is necessary for acceptance of this PR.  Doing that would reduce the number of times the mutex is hit compared to my previous concern:  once per file open versus every read/write operation.  And if some user code accidentally removes a key that is needed, system will have read errors ... so more code needed to make sure there is a union of current read keys with new keys (including forcing the current write key into read list when write is changed).

Reply:
sounds good.

Reply:
Ok to leave as extension in future PR, or Ok as to design and must implement now?